lib/spack/spack/installer.py

```diff
@@ -2420,7 +2420,9 @@ def get_deptypes(self, pkg):
         else:
             cache_only = self.install_args.get("dependencies_cache_only")
 
-        if not cache_only or include_build_deps:
+        # Only add build dependencies if pkg is not already installed
+        # and build depdencies are called for by install_args.
+        if (not cache_only or include_build_deps) and not pkg.spec.installed:
```

Comment:
This looks incorrect to me. `spack install --include-build-deps` should disable build edge pruning. So you probably want

```suggestion
        if include_build_deps or not (cache_only or pkg.spec.installed):
```

So, this pruning optimization only happens when it's not disabled with --include-build-deps, and source builds are allowed and the package isnt installed.